source/server/init_manager_impl.cc

```diff
@@ -4,17 +4,24 @@
 
 #include "common/common/assert.h"
 
+#define TRACE_INIT_MANAGER_WITH_DESCRIPTION(fmt, description, ...)                                 \
+  ENVOY_LOG(debug, "InitManagerImpl({}): " fmt, description, ##__VA_ARGS__)
+
 #define TRACE_INIT_MANAGER(fmt, ...)                                                               \
-  ENVOY_LOG(debug, "InitManagerImpl({}): " fmt, description_, ##__VA_ARGS__)
+  TRACE_INIT_MANAGER_WITH_DESCRIPTION(fmt, description_, ##__VA_ARGS__)
 
 namespace Envoy {
 namespace Server {
 
-InitManagerImpl::InitManagerImpl(absl::string_view description) : description_(description) {
+InitManagerImpl::InitManagerImpl(absl::string_view description)
+    : self_(std::make_shared<InitManagerImpl*>(this)), description_(description) {
   TRACE_INIT_MANAGER("constructor");
 }
 
-InitManagerImpl::~InitManagerImpl() { TRACE_INIT_MANAGER("destructor"); }
+InitManagerImpl::~InitManagerImpl() {
+  TRACE_INIT_MANAGER("destructor");
+  *self_ = nullptr;
```

Comment:
Need to ensure that `InitManager` is non-copyable and non-movable for this to work, but seems like it could resolve the issue.

Another way to do this would be to use `weak_ptr` in the init target but I don't see that as a material improvement.